internal/inputs/generator_data.go

```diff
@@ -0,0 +1,270 @@
+	}
+	return nil
+}
+
+// DataGeneratorConfig is the GeneratorConfig that should be used with a
+// DataGenerator. It includes all the fields from a BaseConfig, as well as some
+// options that are specific to generating the data for database write operations,
+// such as the initial scale and how spaced apart data points should be in time.
+type DataGeneratorConfig struct {
+	BaseConfig
+	InitialScale         uint64
+	LogInterval          time.Duration
+	InterleavedGroupID   uint
+	InterleavedNumGroups uint
+}
+
+func (c *DataGeneratorConfig) Validate() error {
+	err := c.BaseConfig.Validate()
+	if err != nil {
+		return err
+	}
+
+	if c.InitialScale == 0 {
+		c.InitialScale = c.BaseConfig.Scale
+	}
+
+	if c.LogInterval == 0 {
+		return fmt.Errorf(errLogIntervalZero)
+	}
+
+	err = validateGroups(c.InterleavedGroupID, c.InterleavedNumGroups)
+	return err
+}
+
+func (c *DataGeneratorConfig) AddToFlagSet(fs *flag.FlagSet) {
+	c.BaseConfig.AddToFlagSet(fs)
+	flag.Uint64Var(&c.InitialScale, "initial-scale", 0, "Initial scaling variable specific to the use case (e.g., devices in 'devops'). 0 means to use -scale value")
+	flag.DurationVar(&c.LogInterval, "log-interval", defaultLogInterval, "Duration between host data points")
+
+	flag.UintVar(&c.InterleavedGroupID, "interleaved-generation-group-id", 0,
+		"Group (0-indexed) to perform round-robin serialization within. Use this to scale up data generation to multiple processes.")
+	flag.UintVar(&c.InterleavedNumGroups, "interleaved-generation-groups", 1,
+		"The number of round-robin serialization groups. Use this to scale up data generation to multiple processes.")
+
+}
+
+// DataGenerator is a type of Generator for creating data that will be consumed
+// by a database's write/insert operations. The output is specific to the type
+// of database, but is consumed by TSBS loaders like tsbs_load_timescaledb.
+type DataGenerator struct {
+	// Out is the writer where data should be written. If nil, it will be
+	// os.Stdout unless File is specified in the GeneratorConfig passed to
+	// Generate.
+	Out io.Writer
+
+	config  *DataGeneratorConfig
+	tsStart time.Time
+	tsEnd   time.Time
+
+	// bufOut represents the buffered writer that should actually be passed to
+	// any operations that write out data.
+	bufOut *bufio.Writer
+}
+
+func (g *DataGenerator) init(config GeneratorConfig) error {
+	if config == nil {
+		return fmt.Errorf(ErrNoConfig)
+	}
+	switch config.(type) {
+	case *DataGeneratorConfig:
+	default:
+		return fmt.Errorf(ErrInvalidDataConfig)
+	}
+	g.config = config.(*DataGeneratorConfig)
+
+	err := g.config.Validate()
+	if err != nil {
+		return err
+	}
+
+	g.tsStart, err = ParseUTCTime(g.config.TimeStart)
+	if err != nil {
+		return fmt.Errorf(errCannotParseTimeFmt, g.config.TimeStart, err)
+	}
+	g.tsEnd, err = ParseUTCTime(g.config.TimeEnd)
+	if err != nil {
+		return fmt.Errorf(errCannotParseTimeFmt, g.config.TimeEnd, err)
+	}
+
+	if g.Out == nil {
```

Comment:
Potential suggestion (but not really a blocker for merging this): One thing that might be useful here for external tooling is if we could somehow specify an in-memory buffer that implements the io.Writer interface (like bytes.Buffer or any custom writer) instead of just stdout or a file. Then a tool could potentially handle the entire data generation/loading pipeline in memory without having to rely on intermediate artifacts on disk or stdout, which could be useful in some cases. 

This becomes way more compelling for handling the output of queries and inserts, where we could also implement custom readers to help tools digest TSBS output more easily. But could definitely see it being useful for the generation phase as well.

Reply:
I designed this to work a bit like https://golang.org/pkg/os/exec/#Cmd here, so you actually can do that. Basically you would just set `Out` to the `io.Writer` you want it to go to and it should work as expected. This conditional only captures the case where you don't do that, in which case it sets it to os.Stdout.

Example:
```
func main() {
  c := &DataGeneratorConfig{...}
  var buf bytes.Buffer
  dg := &DataGenerator{Out: &buf}  // all output should be written to buf, not Stdout
  dg.Generate(c)
```